webgl: object deletion and bindings

anholt opened the pull request:
<!-- Please describe your changes on the following line: -->

This PR brings in some fixes to the wpt object-deletion-and-behaviour.html testcase.

The test still has many failures, but some of the things that were really getting in the way are fixed.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12934)

<!-- Reviewable:end -->

anholt committed:
webgl: Remove objects from binding points on object deletion.

We keep bindings that shadow what's mapped in the GL state currently,
and so we need to remove the objects from our binding points when they
get implicitly removed from the GL binding points during object deletion.

anholt committed:
webgl: Protect against GL error on glBindRenderbuffer(deleted rbo).

On a GLES or compatibility underlying GL context, we were fine because
an underlying renderbuffer object would just get re-created, and
nothing too bad happened because we aren't tracking the currently
bound renderbuffer at the DOM level.  However, on a desktop GL core
context, binding non-genned or deleted names is an error.

Fixes a crash in object-deletion-behavior.html.

anholt committed:
webgl: Add getters for other GL_*_BINDING enums.

Fixes many subcases of object-deletion-behaviour.html.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webglrenderingcontext.rs
 * @emilio: components/script/dom/webglrenderingcontext.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

anholt commented:
Note that to generate my updates to test expecteds, I'm doing:

- On master, ./mach update-wpt --ignore-existing <wpt results from master>
- Commit results with "DO NOT PUSH" commit message
- Rebase branch on this
- Generate new ./mach update-wpt --ignore-existing <wpt results from branch
- Include these changes in my commit fixing the code.
- Rebase -i branch to drop "DO NOT PUSH" commit, fixing conflicts from dropping the other WPT updates.

Is this a reasonable way to be updating test results?  It seems like we want to be reducing the number of FAILs we'll tolerate, ./mach update-wpt on its own won't do that for the subtests, but ./mach update-wpt --ignore-existing does a *ton* of changes.

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -705,13 +705,16 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
-                .send(CanvasMsg::WebGL(WebGLCommand::BindRenderbuffer(target, None)))
-                .unwrap()
+        match renderbuffer {
+            Some(renderbuffer) if !renderbuffer.is_deleted() => {
```
Good catch! I think this behavior (unbinding even if the renderbuffer is deleted) is tricky. Could you clarify this? A quote to the GLES 2.0 spec like:

> A renderbuffer object binding created with glBindRenderbuffer remains active **until a different renderbuffer object name is bound**, or until the bound renderbuffer object is deleted with glDeleteRenderbuffers.

(emphasis mine).

Would be helpful, assuming that *until a different renderbuffer object is bound* catches the case of naming a deleted renderbuffer? But it seems to me that after calling DeleteRenderbuffer, we shouldn't consider that a valid renderbuffer, so we should silently ignore it, instead of unbinding the current one. This would be quoting the [GLES 2.0 glDeleteRenderbuffers page](https://www.khronos.org/opengles/sdk/docs/man/xhtml/glDeleteRenderbuffers.xml):

> After a renderbuffer object is deleted, it has no contents, and its name is free for reuse[...]

And the WebGL 1.0 spec, that only specifies:

>  If renderbuffer is null the renderbuffer object currently bound to this target is unbound.

In fact, just checked the mesa src, and what they seem to do in `glDeleteRenderbuffers`, is deleting immediately the renderbuffer from the hash table, so when the name is given to `glBindRenderbuffer` it's not found.

In the not found case (and when `allow_user_names` is `false`, that I guess it's on core profile, they generate an `INVALID_OPERATION` error.

So given the `INVALID_OPERATION` error is not spec'd in WebGL nor GLES2, nor unbinding the current renderbuffer, I think the sensitive thing is ignoring it. What do you think?

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -71,6 +71,20 @@ macro_rules! handle_object_deletion {
     };
 }
 
+macro_rules! object_binding_to_js {
```
nit: Maybe calling this `object_binding_to_js_or_null`, even if a bit more annoying, would allow not having to come here to see what the behavior is when it's `None`? It's a uber-nit though, and sort of opinionated, so I wouldn't gate landing this on it.

emilio commented:
The rest of the commits, once the spec thing is clarified look good to me fwiw.

Also, it seems you've stomped for the first time with `tidy`, our style checker! :)

```
./components/script/dom/webglrenderingcontext.rs:547: Line is longer than 120 characters
```

emilio commented:
@anholt: (sorry for not replying before, just read it).

Looks reasonable to update the tests that way. If you want, an easier way to do it until our flakiness is reduced is doing a try run once the PR is open, then updating the tests with the wpt log from the bots.

anholt commented:
Oops. I ran tidy, and totally missed the output (thus the travis fail)

anholt commented:
Spec-wise, deleting an object definitely removes the object from the binding point (even internal binding points, like renderbuffers/textures attached to FBOs.  Yes, this is awful API design.)

We may need to change how we handle reusing our WebGL* objects after deletion (I think that's what some of the remaining failures are about), but that's separate from the unbinding here.

I've tried to do the fixups from review, but the build is taking forever so I'm pushing with ./mach build -d not done yet and going to bed.

emilio commented:
Yeah, my concern is when we call `DeleteRenderbuffer` but there's another renderbuffer currently bound. Anyway, late for me too, so tomorrow will re-read :)

emilio commented:
@bors-servo: try

I still think the current renderbuffer binding shouldn't be deleted when calling `BindRenderbuffer` with a deleted one. Agreed we should unbind the current binding if it's deleted, but you handle that now with your second patch.

My concern is the following:

 * Renderbuffer A is bound.
 * Renderbuffer B is created, and subsequently destroyed. Renderbuffer A is still bound.
 * Call to `BindRenderbuffer(B)`.

With your current code, we unbind A in the third step, and I don't see any spec that would back that behavior (and I think it's quite counter-intuitive).

In compat mode, a new buffer is created, and in core mode, an `INVALID_OPERATION` is thrown. Since none of those behaviors is specified by neither GLES2 or WebGL 1.0, I propose to ignore the third call silently.

bors-servo commented:
:hourglass: Trying commit df86a7c with merge 640cd7c...

emilio commented on components/script/dom/webglrenderingcontext.rs:
```diff
@@ -513,19 +547,20 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
+            constants::RENDERBUFFER_BINDING => return object_binding_to_js_or_null!(cx, &self.bound_renderbuffer),
+            constants::TEXTURE_BINDING_2D => return object_binding_to_js_or_null!(cx, &self.bound_texture_2d),
+            constants::TEXTURE_BINDING_CUBE_MAP => {
+                return object_binding_to_js_or_null!(cx, &self.bound_texture_cube_map);
```
If the braces are just due to a line length issue, why not formatting it like this to be consistent?

```rust
constants::ARRAY_BUFFER_BINDING
    => return object_binding_to_js_or_null!(cx, &self.bound_buffer_array),
constants::ELEMENT_ARRAY_BUFFER_BINDING
    => return object_binding_to_js_or_null!(cx, &self.bound_buffer_element_array),
constants::FRAMEBUFFER_BINDING
    => return object_binding_to_js_or_null!(cx, &self.bound_framebuffer),
// ...
```

emilio commented:
@bors-servo: try

 * Don't know where did the previous try run go.

bors-servo commented:
:hourglass: Trying commit ed517b9 with merge b4d4eff...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2720)

emilio commented:
@bors-servo: r+

Sorry for the delay in getting this reviewed :/

bors-servo commented:
:pushpin: Commit 2433f70 has been approved by `emilio`

<!-- @bors-servo r=emilio 2433f703776e9f13ac6b900f62a8870341f57957 -->

emilio commented:
@bors-servo: r-

Actually, I thought the first commit was gone. Do you still think that's right @anholt? If you could point to a spec that backs that out I'd be glad to r+, except that `BindRenderbuffer` discussion the rest looks great to me.

anholt commented:
Some samples from the GL 4.4 core spec:

"When a buffer, texture, renderbuffer, query, transform feedback, or sync object is
deleted, its name immediately becomes invalid (e.g. is marked unused), but the
underlying object will not be deleted until it is no longer in use."

"Caution should be taken when deleting an object attached to a container ob-
ject (such as a buffer object attached to a vertex array object, or a renderbuffer
or texture attached to a framebuffer object), or a shared object bound in multiple
contexts. Following its deletion, the object’s name may be returned by Gen* com-
mands, even though the underlying object state and data may still be referred to by
container objects, or in use by contexts other than the one in which the object was
deleted."

"A framebuffer object is created by binding a name returned by GenFrame-
buffers (see below) to DRAW_FRAMEBUFFER or READ_FRAMEBUFFER ."

This seems really clear to me: When you've deleted the name, you can't use that name any more.  You need to re-gen to get a new name.

emilio commented:
@anholt: Yeah, I see that, what I don't see is why you need to unbind when the current bound buffer is other.

What i mean is.

emilio commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit e8b5f70 has been approved by `emilio`

<!-- @bors-servo r=emilio e8b5f70429b70e779e959f616fc88456a52bc7a4 -->

bors-servo commented:
:hourglass: Testing commit e8b5f70 with merge 969aefe...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2672), [arm64](http://build.servo.org/builders/arm64/builds/2688), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2741), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2903), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2754), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2743), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2843), [windows-dev](http://build.servo.org/builders/windows-dev/builds/637)
